CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+#include "DesktopNotificationManagerCompat.h"
+#include <appmodel.h>
+#include <string>
+
+using namespace Microsoft::WRL;
+using namespace Microsoft::WRL::Wrappers;
+
+bool DesktopNotificationManagerCompat::s_registeredAumidAndComServer = false;
+const wchar_t *DesktopNotificationManagerCompat::s_aumid = nullptr;
+bool DesktopNotificationManagerCompat::s_registeredActivator = false;
+bool DesktopNotificationManagerCompat::s_hasCheckedIsRunningAsUwp = false;
+bool DesktopNotificationManagerCompat::s_isRunningAsUwp = false;
+
+HRESULT DesktopNotificationManagerCompat::RegisterAumidAndComServer(const wchar_t *aumid, GUID clsid)
+{
+    // If running as Desktop Bridge
+    if (IsRunningAsUwp())
+    {
+        // Clear the AUMID since Desktop Bridge doesn't use it, and then we're done.
+        // Desktop Bridge apps are registered with platform through their manifest.
+        // Their LocalServer32 key is also registered through their manifest.
+        s_aumid = nullptr;
+        s_registeredAumidAndComServer = true;
+        return S_OK;
+    }
+
+    s_aumid = aumid;
+
+    // Get the EXE path
+    wchar_t exePath[MAX_PATH];
+    DWORD charWritten = ::GetModuleFileName(nullptr, exePath, ARRAYSIZE(exePath));
+    auto hr = charWritten > 0 ? S_OK : HRESULT_FROM_WIN32(::GetLastError());
+    if (SUCCEEDED(hr))
+    {
+        // Register the COM server
+        hr = RegisterComServer(clsid, exePath);
+        if (SUCCEEDED(hr))
+        {
+            s_registeredAumidAndComServer = true;
+        }
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterActivator()
+{
+    // Module<OutOfProc> needs a callback registered before it can be used.
+    // Since we don't care about when it shuts down, we'll pass an empty lambda here.
+    Module<OutOfProc>::Create([] {});
+
+    // If a local server process only hosts the COM object then COM expects
+    // the COM server host to shutdown when the references drop to zero.
+    // Since the user might still be using the program after activating the notification,
+    // we don't want to shutdown immediately.  Incrementing the object count tells COM that
+    // we aren't done yet.
+    Module<OutOfProc>::GetModule().IncrementObjectCount();
+
+    auto hr = Module<OutOfProc>::GetModule().RegisterObjects();
+    if (SUCCEEDED(hr))
+    {
+        s_registeredActivator = true;
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterComServer(GUID clsid, wchar_t exePath[])
+{
+    // Turn the GUID into a string
+    OLECHAR* clsidOlechar;
+    StringFromCLSID(clsid, &clsidOlechar);
+    std::wstring clsidStr(clsidOlechar);
+    ::CoTaskMemFree(clsidOlechar);
```

Comment:
If line 73 throws this thing leaks. Would prefer clsidOlechar to be RAII

Reply:
Could you provide an example of what the code should be? Seems like that's a reference to object owning resources, but my limited C++ experience likely isn't going to lead to the right answer :)